File: src/test/scala/de/uni_mannheim/desq/mining/spark/DesqMiningTest.scala
package de.uni_mannheim.desq.mining.spark

import java.io.{File, FileOutputStream}

import de.uni_mannheim.desq.io.DelPatternWriter
import de.uni_mannheim.desq.mining.WeightedSequence
import de.uni_mannheim.desq.util.{DesqProperties, TestUtils}
import org.apache.log4j.Logger
import org.apache.spark.SparkContext
import org.assertj.core.api.Assertions._
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.scalatest.junit.AssertionsForJUnit

import scala.reflect.ClassTag

/**
  * Created by rgemulla on 20.09.2016.
  */
@RunWith(classOf[Parameterized])
abstract class DesqMiningTest(sigma: Long, patternExpression: String,
                              minerName: String, conf: DesqProperties) extends AssertionsForJUnit {
  /** The data */
  def getDesqDataset()(implicit sc: SparkContext): DesqDataset
  def getGenericDesqDataset()(implicit sc: SparkContext): GenericDesqDataset[(Array[String], Long)]

  def goldFileBaseName: String

  def testDirectoryName: String


  @Test
  def testDesqDataset() {
    val fileName = goldFileBaseName+ "-" + sigma + "-" + sanitize(patternExpression) + ".del"
    val actualFile = TestUtils.newTemporaryFile(
      TestUtils.getPackageResourcesPath(getClass) + "/" + testDirectoryName + "/" + minerName + "/" + fileName)
    mineWithDesqDataset(actualFile)
    try {
      val expectedFile = TestUtils.getPackageResource(classOf[de.uni_mannheim.desq.mining.DesqMiningTest], fileName) // use path of sequential results
      assertThat(actualFile).hasSameContentAs(expectedFile)
    } catch {
      case e: NullPointerException =>
        DesqMiningTest.logger.error("Can't access expected data file for " + actualFile)
        throw e
    }
  }

  @Test
  def testGenericDesqDataset() {
    val fileName = goldFileBaseName+ "-" + sigma + "-" + sanitize(patternExpression) + ".del"
    val actualFile = TestUtils.newTemporaryFile(
      TestUtils.getPackageResourcesPath(getClass) + "/" + testDirectoryName + "/" + minerName + "/" + fileName)
    mineWithGenericDesqDataset(actualFile)
    try {
      val expectedFile = TestUtils.getPackageResource(classOf[de.uni_mannheim.desq.mining.DesqMiningTest], fileName) // use path of sequential results
      assertThat(actualFile).hasSameContentAs(expectedFile)
    } catch {
      case e: NullPointerException =>
        DesqMiningTest.logger.error("Can't access expected data file for " + actualFile)
        throw e
    }
  }

  def mineWithDesqDataset(outputDelFile: File) {
    implicit val sc = de.uni_mannheim.desq.util.spark.TestUtils.sc

    val data = getDesqDataset()

    // Perform pattern mining into del file
    val resultRDD = data.mine(conf)
    val result = resultRDD.sequences.collect()

    // write the data
    val patternWriter = new DelPatternWriter(new FileOutputStream(outputDelFile), DelPatternWriter.TYPE.GID)
    patternWriter.setDictionary(data.descriptor.getDictionary)
    result.foreach(patternWriter.write)
    patternWriter.close()

    // sort del file
    TestUtils.sortDelPatternFile(outputDelFile)
  }

  def mineWithGenericDesqDataset(outputDelFile: File) {
    implicit val sc = de.uni_mannheim.desq.util.spark.TestUtils.sc

    val data = getGenericDesqDataset()

    // Perform pattern mining into del file
    val resultRDD = data.mine(conf)
    val result = resultRDD.sequences.collect()

    // write the data
    val patternWriter = new DelPatternWriter(new FileOutputStream(outputDelFile), DelPatternWriter.TYPE.GID)
    patternWriter.setDictionary(data.descriptor.getDictionary)
    result.foreach(patternWriter.write)
    patternWriter.close()

    // sort del file
    TestUtils.sortDelPatternFile(outputDelFile)
  }

  def sanitize(s: String) = de.uni_mannheim.desq.mining.DesqMiningTest.sanitize(s)
}

object DesqMiningTest {
  private val logger = Logger.getLogger(classOf[DesqMiningTest])
}

File: src/test/java/de/uni_mannheim/desq/mining/DesqMiningTest.java
package de.uni_mannheim.desq.mining;

import de.uni_mannheim.desq.dictionary.Dictionary;
import de.uni_mannheim.desq.io.DelPatternWriter;
import de.uni_mannheim.desq.io.SequenceReader;
import de.uni_mannheim.desq.util.DesqProperties;
import de.uni_mannheim.desq.util.TestUtils;
import org.apache.log4j.Logger;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;

import static org.assertj.core.api.Assertions.assertThat;

/** A test for traditional frequent sequence mining. Datasets and parameters are set by implementing classes.
 *
 * Created by rgemulla on 18.07.2016.
 */
@RunWith(Parameterized.class)
public abstract class DesqMiningTest {
    private static final Logger logger = Logger.getLogger(DesqMiningTest.class);

    long sigma;
    String patternExpression;
    String minerName;
    DesqProperties conf;

    DesqMiningTest(long sigma, String patternExpression, String minerName, DesqProperties conf) {
        this.sigma = sigma;
        this.patternExpression = patternExpression;
        this.minerName = minerName;
        this.conf = conf;
    }

    /** The dictionary to use */
    public abstract Dictionary getDictionary() throws IOException;

    /** Whether counts and fids need to be computed from data */
    public abstract boolean computeStatisticsAndFids();

    /** Reader for data */
    public abstract SequenceReader getSequenceReader() throws IOException;

    public abstract String getGoldFileBaseName();

    public abstract String getTestDirectoryName();

    @Test
    public void test() throws IOException,
            InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
        String fileName = getGoldFileBaseName() + "-" + sigma + "-" + sanitize(patternExpression) + ".del";
        File actualFile = TestUtils.newTemporaryFile(
                TestUtils.getPackageResourcesPath(getClass()) + "/" + getTestDirectoryName() + "/" + minerName + "/" + fileName);
        mine(actualFile);
        try {
            File expectedFile = TestUtils.getPackageResource(getClass(), fileName);
            assertThat(actualFile).hasSameContentAs(expectedFile);
        } catch (NullPointerException e) {
            logger.error("Can't access expected data file for " + actualFile);
            throw e;
        }
    }

    public void mine(File outputDelFile)
            throws IOException, NoSuchMethodException, IllegalAccessException, InvocationTargetException,
                InstantiationException {
        // load the dictionary
        Dictionary dict = getDictionary();

        // update hierarchy
        if (computeStatisticsAndFids()) {
            SequenceReader dataReader = getSequenceReader();
            dataReader.setDictionary(dict);
            dict.clearFreqs();
            dict.incFreqs(dataReader);
            dataReader.close();
            dict.recomputeFids();
        }

        // Perform pattern mining into del file
        SequenceReader dataReader = getSequenceReader();
        dataReader.setDictionary(dict);
        DesqMinerContext ctx = new DesqMinerContext();
        ctx.dict = dict;
        DelPatternWriter patternWriter = new DelPatternWriter(new FileOutputStream(outputDelFile), DelPatternWriter.TYPE.GID);
        patternWriter.setDictionary(dict);
        ctx.patternWriter = patternWriter;
        ctx.conf = conf;
        DesqMiner miner = DesqMiner.create(ctx);
        miner.addInputSequences(dataReader);
        dataReader.close();
        miner.mine();
        patternWriter.close();

        // sort del file
        TestUtils.sortDelPatternFile(outputDelFile);
    }

    /** Sanitizes the given string (often a pattern expression) so that it can be used as filename */
    public static String sanitize(String s) {
        s = s.replace(" ", "-");
        s = s.replace("|", "_I");
        s = s.replace("*", "_S");
        s = s.replace("?", "_Q");
        s = s.replace("[", "_(");
        s = s.replace("]", "_)");
        s = s.replace("^", "_G");
        s = s.replace("=", "_E)");
        return s;
    }
}
